feat(ivy): provide support for map-based host bindings for [style] and [class]

matsko opened the pull request:
Up until now, `[style]` and `[class]` bindings (the map-based ones) have only
worked as template bindings and have not been supported at all inside of host
bindings. This patch ensures that multiple host binding sources (components and
directives) all properly assign style values and merge them correctly in terms
of priority.

Jira: FW-882

trotyl commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+  const oldKeys = Object.keys(a);
+  const newKeys = keys;
+
+  // they keys are different which means the map changed
```
```suggestion
  // the keys are different which means the map changed
```

trotyl commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+  if (a && b) {
+    if (a.length !== b.length) return false;
+    for (let i = 0; i < a.length; i++) {
+      if (b.indexOf(a[i]) === -1) return false;
```
Is O(n^2) necessary?

matsko replied:
Yes because there's no guarantee the arrays are in order.

trotyl replied:
How about using a temporary `Set` to store the checked keys? It has `O(1)` look-up.

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -522,170 +614,302 @@ export function updateStylingMap(
+ * value in order. This means that a directive like `classADirective` will always fire before
+ * `classBDirective` and therefore its styling values (classes and styles) will always be evaluated
+ * in the same order. Because of this consistent ordering, the first directive has a higher priority
+ * than the second one. Its with this prioritzation mechanism that the styling algorithm knows how
```
```suggestion
 * than the second one. With this prioritization mechanism that the styling algorithm knows how
```

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -522,170 +614,302 @@ export function updateStylingMap(
+      context[entryIsClassBased ? StylingIndex.CachedMultiClasses : StylingIndex.CachedMultiStyles];
+
+  // this is the index in which this directive has ownership access to write to this
+  // value (anything before is owned by a previous directive that is more impotant)
```
```suggestion
  // value (anything before is owned by a previous directive that is more important)
```

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -522,170 +614,302 @@ export function updateStylingMap(
+            }
+
+            if (hasValueChanged(distantCtxFlag, distantCtxValue, value)) {
+              if (value == null || value == undefined && value != distantCtxValue) {
```
`value == null` is exactly `value == undefined`, do you mean `===`?

matsko commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+  setValue(value: {[key: string]: any}|string[]|string|null) {
+    if (Array.isArray(value)) {
+      this._lastSetValueType = StylingDifferValueTypes.Array;
+    } else if (value instanceof Set) {
```
Not sure I follow what you mean here.

trotyl replied:
Sorry for not making it clear, I mean `Array.isArray()` will return `true` for an `Array` in different realm, but `instanceof` will return `false`, so this seems to be inconsistency. But given it's used in previous implementation, should not be a breaking change. (Also I think it should allow any `Iterable` besides `Set`)

matsko commented on packages/common/src/directives/styling_differ.ts:
```diff
@@ -0,0 +1,289 @@
+    map: {[key: string]: any}, key: string, value: any, parseOutUnits: boolean,
+    allowSubKeys: boolean) {
+  if (allowSubKeys && key.indexOf(' ') > 0) {
+    const innerKeys = key.split(/\s+/g);
```
Not sure I follow what you mean here.

trotyl replied:
In this logic:

+ `a b` will be considered multi-keys and results to `['a', 'b']`;
+ `a\tb` will **NOT** be considered multi-keys and results to `['a\tb']`;
+ `a b\tc` will be considered multi-keys and results to `['a', 'b', 'c']`;

So I think the space testing and space splitting should follow same rule.

matsko commented:
Thank you @trotyl for your helpful responses.

JoostK commented on packages/common/src/directives/ng_class_base.ts:
```diff
@@ -0,0 +1,184 @@
+import {StylingDiffer, StylingDifferOptions} from './styling_differ';
+
+@Directive({selector: '[ngClass]'})
+export class NgClass__POST_R3__ implements DoCheck {
```
Can we verify somehow that this symbol (and `NgStyle__POST_R3__`) does indeed end up in the final bundle after Rollup? We have seen issues with `__POST_R3__` symbols being dropped because of dead code elimination, if the symbol is not privately exported. I'm not sure if the reference to this symbol from `ɵNgClassR3Module` is enough to keep it alive, as `ɵNgClassR3Module` might be considered dead as well.

matsko committed:
feat(ivy): provide support for map-based host bindings for [style] and [class]

Up until now, `[style]` and `[class]` bindings (the map-based ones) have only
worked as template bindings and have not been supported at all inside of host
bindings. This patch ensures that multiple host binding sources (components and
directives) all properly assign style values and merge them correctly in terms
of priority.

Jira: FW-882

trotyl commented on packages/core/src/render3/interfaces/styling.ts:
```diff
@@ -441,29 +522,94 @@ export const enum SinglePropOffsetValuesIndex {
+ *    values (set by another directive via its `[style]` and/or `[class]` host binding) will not be
+ *    able to set it. This is because the former directive has priorty.
+ * 2. Only if a former directive has set a specific styling value to null (whether by actually
+ *    setting it to null or not including it in is map value) then a less imporatant directive can
```
```suggestion
 *    setting it to null or not including it in its map value) then a less important directive can
```

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -522,170 +618,304 @@ export function updateStylingMap(
+ * than the second one. It is with this prioritzation mechanism that the styling algorithm knows how
+ * to merge and apply redudant styling properties.
+ *
+ * The function itself applies the key/value entries (or a array of keys) to
```
```suggestion
 * The function itself applies the key/value entries (or an array of keys) to
```

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -522,170 +618,304 @@ export function updateStylingMap(
+ *
+ * STEP 1:
+ *    First check to see what properties are already set and in use by another directive in the
+ *    context (e.g. `ngClass` set the `width` value and not `[style.width]="w"` in a directive is
```
```suggestion
 *    context (e.g. `ngClass` set the `width` value and `[style.width]="w"` in a directive is
```

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1401,6 +1636,12 @@ function getDirectiveIndexFromRegistry(context: StylingContext, directive: any)
+function isMapBasedDirectiveDirty(directiveIndex: number): boolean {
+  const index = MapBasedOffsetValuesIndex.ValuesStartPosition +
+      directiveIndex * MapBasedOffsetValuesIndex.Size;
+  return false;
```
Does this really work? Or is it used?

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1520,24 +1761,184 @@ function allowValueChange(
+         i += InitialStylingValuesIndex.Size) {
+      const value = initialStyleValues[i + 1];
+      if (value !== null) {
+        styleString += (styleString.length ? ';' : '') + `${initialStyleValues[i]}:${value}`;
```
Would it better for generate style attributes to always end with `;`? Like `width:100px;height:200px;`? I think that might be more recommended in [style attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/style) usage.

trotyl commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1520,24 +1761,184 @@ function allowValueChange(
+ * - The dirty flag will be set to true
+ *
+ * If the `dirtyFutureValues` param is provided then it will update all future entries (binding
+ * entries that exist as apart of other directives) to be dirty as well. This will force the
```
```suggestion
 * entries that exist as a part of other directives) to be dirty as well. This will force the
```

mhevery commented on packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts:
```diff
@@ -1062,6 +1062,84 @@ describe('compiler compliance: styling', () => {
+                @Component({
+                  selector: 'my-component',
+                  template: \`
+                    <div [style!override]="myStyleExp"
```
Why is this called `override`? In CSS terminology we have `!important` Why not follow?

mhevery commented on packages/core/src/render3/instructions.ts:
```diff
@@ -1405,9 +1430,33 @@ export function elementStyling(
+
+    const fns = tNode.creationInstructionFns = tNode.creationInstructionFns || [];
+    fns.push(
+        () => initElementStyling(
```
This is created a closure here. It seems to be the only closure which ever gets passed into the array. Why not pass only arguments and then just invoke it directly? Is it for tree shaking?

mhevery commented on packages/core/src/render3/interfaces/node.ts:
```diff
@@ -345,6 +348,8 @@ export interface TNode {
    */
   projection: (TNode|RNode[])[]|number|null;
+
+  creationInstructionFns: Function[]|null;
```
Could we have a more descriptive name here? (also missing comment).

mhevery commented on packages/core/src/render3/interfaces/styling.ts:
```diff
@@ -44,9 +44,9 @@ import {PlayerContext} from './player';
- * the styling context is created then single and multi properties can stored within it. For this to
- * happen, the following function needs to be called:
+ * In either case, a styling context will be created and stored within an element's `LViewData`.
+ * Once the styling context is created then single and multi properties can stored within it. For
```
```suggestion
 * Once the styling context is created then single and multi properties can be stored within it. For
```

mhevery commented on packages/core/src/render3/interfaces/styling.ts:
```diff
@@ -187,18 +206,57 @@ import {PlayerContext} from './player';
+ * then the algorithm will decide how to update the value based on the following heuristic:
+ *
+ * 1. If the template binding has a value then it always wins
+ * 2. If not then whichever first-registered directive that has that value first will win
```
```suggestion
 * 2. Otherwise whichever first-registered directive that has that value first will win
```

mhevery reviewed (commented):
Overall it looks good. Left few minor comments.

matsko commented on packages/core/src/render3/styling/class_and_style_bindings.ts:
```diff
@@ -1520,24 +1761,184 @@ function allowValueChange(
+
+function hyphenate(value: string): string {
+  return value.replace(
+      /[a-z][A-Z]/g, match => `${match.charAt(0)}-${match.charAt(1).toLowerCase()}`);
```
it needs to be inside of `render3`

matsko committed:
fixup! feat(ivy): provide support for map-based host bindings for [style] and [class]
